### ISSUE TITLE
[area/harvester] failure to provision with air-gapped environment

### ISSUE THREAD

**Issue by dweomer** [rancher/rancher#36488]
**Rancher Server Setup**
- Rancher version:
  - **2.6.3**
- Installation option (Docker install/Helm Chart):
   - **n/a**
- Proxy/Cert Details:
  - **n/a**

**Information about the Cluster**
- Kubernetes version:
  - **1.22.x**
  - **1.21.x**
- Cluster Type (Local/Downstream):
   - **Downstream :arrow_right: Harvester 1.0.0**

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - **admin**

   
**Describe the bug**
- https://github.com/harvester/harvester/issues/1808

> **Expected behavior**
> 
> The driver binaries of "built-in" node drivers are included in the Rancher release and do not have to be downloaded post-install. 
> 
> This is the case for other node drivers of type "built-in", such as AWS, Azure, etc.  The `url` field of these node drivers  is set to `"url": "local://",`
> 
> **Actual behavior**
> 
> In Rancher 2.6.3 the Harvester node driver is marked as "built-in" driver, yet the `url` field is set to a external URL `https://releases.rancher.com/harvester-node-driver/v0.3.4/docker-machine-driver-harvester-amd64.tar.gz`.
> 
> When provisioning a Harvester RKE2 cluster from Rancher, the provisioning will fail on air-gapped Rancher instances since the driver can't be downloaded from the internet.
> 

**Screenshots**
![image](https://user-images.githubusercontent.com/1033444/153657755-8f3211e8-a25b-487d-a156-4eafd5134df3.png)

**Additional context**
- https://github.com/harvester/harvester/issues/1808

**Comment by thedadams** [comment]
### Root cause
rancher-machine downloads non-built-in machine drivers before creating machines. The problem in air-gapped environments is that the rancher-machine pod doesn't have access to the internet.

### What was fixed, or what changes have occurred
Now, when node drivers are downloaded by Rancher, they are also put in a directory that Rancher serves via HTTP. Since the pod can communicate with the Rancher server (it is in the Rancher cluster), then rancher-machine will download the driver from Rancher instead.

### Areas or cases that should be tested
1. Provisioning Harvester RKE2 clusters in air-gapped environment.
2. Provisioning another RKE2 cluster with non-built-in node driver.

### What areas could experience regressions?
Number 2 above

### Are the repro steps accurate/minimal?
Yes

**Comment by noahgildersleeve** [comment]
I was unable to import the Harvester cluster into Rancher while in an airgapped environment. It just stayed in pending. I verified that if I added back in internet connectivity it could import fine.

**Comment by thedadams** [comment]
@noahgildersleeve This issue is not about importing a Harvester cluster into Rancher. As far as I know, that was working and I have never worked on that functionality.

This issue was about creating an RKE2 cluster in Harvester in an air-gapped environment.

Could you open a new issue for the problem you are seeing?

**Comment by TachunLin** [comment]
Verified **fixed** on `master-fd872c99-head` (03/09) with Rancher `v2.6.4-rc6`.

### Result
Can provision Harvester RKE2 cluster in Rancher under air gapped environment with proxy server. 
![image](https://user-images.githubusercontent.com/29251855/157668972-32eaf3bb-a980-4b52-b208-f3449e536b6b.png)

![image](https://user-images.githubusercontent.com/29251855/157679857-ba3fb6cf-0d66-4f34-b540-43793f372e1a.png)

### Test Information
* Test Environment: 1 nodes harvester on local kvm machine
* Harvester version: `master-fd872c99-head` (03/09)
* Rancher version: v2.6.4-rc6
* Kubernetes version: 1.21.10+rke2r1
* OS image: Ubuntu cloud focal

### Verify Steps
1. Clone ipxe example 
2. Enable `offline` feature in settings.yaml
3. Use vagrant pxe example to build an one node harvester
4. Open http proxy settings in harvester 
5. Add HTTP and HTTPS proxy setting
  ![image](https://user-images.githubusercontent.com/29251855/157666064-7b6a814f-965c-4767-a510-e64ccf85e2b6.png)
6. Prepare a ubuntu VM and config virtual network `harvester` provided by pxe example
7. Pull rancher v2.6.4-rc6 docker image 
8. Run docker command to create rancher with http proxy arguments 
  ```
  sudo docker run -d --restart=unless-stopped -p 80:80 -p 443:443 \
  -e HTTP_PROXY="http://192.168.0.254:3128" \
  -e HTTPS_PROXY="http://192.168.0.254:3128" \
  -e NO_PROXY="localhost,127.0.0.1,0.0.0.0,10.0.0.0/8,cattle-system.svc,192.168.0.0/16,.svc,.cluster.local,example.com" \
  --privileged rancher/rancher:v2.6.4-rc6
  ```
9.  Import harvester to rancher dashboard 
10. Create cloud credential in rancher
11. Enable virtual network in harvester with `harvester-mgmt` 
12. Create the virtual network vlan1 with id 1 
13. Create a RKE2 cluster 
  ![image](https://user-images.githubusercontent.com/29251855/157666567-d7d3ac1d-c314-4614-aaae-e87776ab0daa.png)
14. Add the following content in User data under `Advanced` 
  ```
  #cloud-config
  apt:
    http_proxy: http://192.168.0.254:3128
    https_proxy: http://192.168.0.254:3128
  password: 123456
  chpasswd: { expire: False }
  ssh_pwauth: True
  ```
  
  ![image](https://user-images.githubusercontent.com/29251855/157681940-273bf456-4dfd-4f56-95a6-f0607cd335a5.png)
  
15. Open `Agent Environment Vars` 
16. Add the following variable and value pair 
```
HTTP_PROXY: http://192.168.0.254:3128
HTTPS_PROXY: http://192.168.0.254:3128
NO_PROXY: localhost,127.0.0.1,0.0.0.0,10.0.0.0/8,cattle-system.svc,.svc,.cluster.local

```
  ![image](https://user-images.githubusercontent.com/29251855/157666993-7e16de1c-5143-45e3-9dc6-4ad7d9ed10aa.png)
  ![image](https://user-images.githubusercontent.com/29251855/157667232-f2547854-74cb-48d8-ad66-23d7f68de7ea.png)
17. Click the create button


### Support Bundle
[supportbundle_51f93467-f4f6-4eca-aeaa-0727d415787f_2022-03-10T13-14-38Z.zip](https://github.com/harvester/harvester/files/8223668/supportbundle_51f93467-f4f6-4eca-aeaa-0727d415787f_2022-03-10T13-14-38Z.zip)

**Comment by slickwarren** [comment]
the validation above is using a proxy. Airgap refers to a setup with no access to the internet, even through proxy. You need to have a registry setup with all images for rancher and harvester in it in order to test this in airgap. cc @noahgildersleeve can provide more information for the harvester team on how to achieve this.

**Comment by noahgildersleeve** [comment]
Testing blocked by https://github.com/harvester/harvester/issues/2046.

**Comment by TachunLin** [comment]
### Update the import Harvester to Rancher status

Now can import Harvester to Rancher when both of them are under Airgapped environment (with no proxy server)
![image](https://user-images.githubusercontent.com/29251855/160383219-4e13cd46-35f4-4a79-acd6-ccb9bcf87662.png)

* The airgapped environment setup include the following phases: 
Please check the **Environment setup** section for details

1. Prepare the air gapped harvester 
2. Create a virtual machine for Rancher in air gapped environment 
3. Setup a private docker registry and download all the offline images for Rancher 
4. Push the image to docker registry 
5. Install K3s and Rancher 
6. Edit Rancher vm hosts file and boostrap Rancher
7. Edit Harvester cluster hosts files
8. Add host mapping to coredns configmap and deployments 
9. Import Harvester from Rancher UI

### Test Information
* Test Environment: 1 node harvester on local kvm machine
* Harvester version: v1.0.1-rc1
* Rancher version: v2.6.4-rc13

### Environment setup

#### Phase 1: setup airgap Harvester
1. Set offline: true in [vagrant-pxe-harvester](https://github.com/harvester/ipxe-examples/blob/f75483563f192090dadd48051bbc3b538c30cd34/vagrant-pxe-harvester/settings.yml#L42-L44).
2. Set 1 node harvester with sufficient resources
3. Run ./setup_harvester.sh.

#### Phase 2: Create a virtual machine for Rancher in air gapped environment 
1. Create a virtual machine with at least 300GB storage (OS: Ubuntu Desktop 20.04)
2. Add `harvester` and `vagrant-libvirt` network to the virtual machine  
  - `harvester` for internal 
  - `vagrant-libvirt` for external 

#### Phase 3: Setup a private docker registry and download all the offline images for Rancher 
1. Install VIM
  ```
  sudo apt update
  sudo apt install vim
  ```
2. Install Docker 
  ```
  sudo apt-get update
  sudo apt-get install \
      ca-certificates \
      curl \
      gnupg \
      lsb-release
  curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg
  echo \
    "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu \
    $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
  sudo apt-get update
  sudo apt-get install docker-ce docker-ce-cli containerd.io
  ```
3. Add current user to docker group
  ```
  sudo groupadd docker
  sudo usermod -aG docker $USER
  ```
4. Logout and login again.
5. Install helm
  ```
  curl https://baltocdn.com/helm/signing.asc | sudo apt-key add -
  sudo apt-get install apt-transport-https --yes
  echo "deb https://baltocdn.com/helm/stable/debian/ all main" | sudo tee /etc/apt/sources.list.d/helm-stable-debian.list
  sudo apt-get update
  sudo apt-get install helm
  ```
6. Create certs folder
  ```
  mkdir -p certs
  ```
7. Generate private registry certificate files
  ```
  openssl req \
    -newkey rsa:4096 -nodes -sha256 -keyout certs/domain.key \
    -addext "subjectAltName = DNS:myregistry.local" \
    -x509 -days 365 -out certs/domain.crt
  ```
8. Move certificate files
  ```
  sudo mkdir -p /etc/docker/certs.d/myregistry.local:5000
  sudo cp certs/domain.crt /etc/docker/certs.d/myregistry.local:5000/domain.crt
  ```
9. Start docker registry
  ```
  docker run -d \
    -p 5000:5000 \
    --restart=always \
    --name registry \
    -v "$(pwd)"/certs:/certs \
    -v "$(pwd)"/registry:/var/lib/registry \
    -e REGISTRY_HTTP_TLS_CERTIFICATE=/certs/domain.crt \
    -e REGISTRY_HTTP_TLS_KEY=/certs/domain.key \
    registry:2
  ```
10. Add myregistry.local record. Remember to change your private ip.
  ```
  # vim /etc/hosts
  192.168.0.50 myregistry.local
  ```
11. Create `get-rancher-scripts` script.
  ```
  # vim get-rancher-scripts
  #!/bin/bash
  if [[ $# -eq 0 ]] ; then
      echo 'This requires you to pass a version for the url like "v2.6.3"'
      exit 1
  fi
  wget https://github.com/rancher/rancher/releases/download/$1/rancher-images.txt
  wget https://github.com/rancher/rancher/releases/download/$1/rancher-load-images.sh
  wget https://github.com/rancher/rancher/releases/download/$1/rancher-save-images.sh
  chmod +x ./rancher-save-images.sh
  chmod +x ./rancher-load-images.sh
  ```
12. Make `get-rancher-scripts` can be executed.
  ```
  chmod +x get-rancher-scripts
  ```
13. Get  rancher-images.txt.
  ```
  ./get-rancher-scripts v2.6.4-rc13
  ```
14. Add cert-manager images to rancher-images.txt
  ```
  helm repo add jetstack https://charts.jetstack.io/
  helm repo update
  helm fetch jetstack/cert-manager --version v1.7.1
  helm template ./cert-manager-v1.7.1.tgz | awk '$1 ~ /image:/ {print $2}' | sed s/\"//g >> ./rancher-images.txt
  ```
15. Sort rancher-images.txt
  ```
  sort -u rancher-images.txt -o rancher-images.txt
  ```
16. Get Rancher images. This step may take at least 2 hours depends on your network speed
  ```
  ./rancher-save-images.sh --image-list ./rancher-images.txt
  ```
  ![image](https://user-images.githubusercontent.com/29251855/160348531-f6f5dd3b-af01-43e2-81ae-ab5b47719d7f.png)
  
#### Phase 4: Push the image to docker registry 
1. Download k3s v1.23.4+k3s1
  ```
  wget https://github.com/k3s-io/k3s/releases/download/v1.23.4%2Bk3s1/k3s-airgap-images-amd64.tar
  wget https://github.com/k3s-io/k3s/releases/download/v1.23.4%2Bk3s1/k3s
  chmod +x k3s
  
  sudo cp k3s /usr/local/bin/
  sudo chown $USER /usr/local/bin/k3s
  
  curl https://get.k3s.io/ -o install.sh
  chmod +x install.sh
  ```
2. Download rancher v2.6,4-rc13 helm
  ```
  helm repo add rancher-latest https://releases.rancher.com/server-charts/latest
  helm fetch rancher-latest/rancher --version=v2.6.4-rc13
  ```
3. Download cert-manager crds.
  ```
  mkdir cert-manager
  curl -L -o cert-manager/cert-manager-crd.yaml https://github.com/jetstack/cert-manager/releases/download/v1.7.1/cert-manager.crds.yaml
  ```

4. Install k9s (option)
  ```
  curl -kL https://github.com/derailed/k9s/releases/download/v0.25.18/k9s_Linux_x86_64.tar.gz > k9s.tar.gz
  tar -zxvf k9s.tar.gz
  mv k9s /usr/local/bin/
  ```

5. Cut Network -> Remove the `vagrant-libvirt` network device from virtual machine, make it air gapped
6. Load Rancher images to private registry. 
  ```
  ./rancher-load-images.sh --image-list ./rancher-images.txt --registry myregistry.local:5000
  ```
  - This step may take 50 minutes depends on total image size

#### Phase 5: Install K3s and Rancher 
1. Move k3s images files.
  ```
  sudo mkdir -p /var/lib/rancher/k3s/agent/images/
  sudo cp ./k3s-airgap-images-amd64.tar /var/lib/rancher/k3s/agent/images/
  ```
2. Install k3s
  ```
  INSTALL_K3S_SKIP_DOWNLOAD=true ./install.sh
  ```
3. Add `registries.yaml`  to `/etc/rancher/k3s/`
  ```
  # vim /etc/rancher/k3s/registries.yaml
  mirrors:
    docker.io:
      endpoint:
        - "https://myregistry.local:5000/"
  configs:
    "myregistry.local:5000":
      tls:
        insecure_skip_verify: true
  ```
4. Restart k3s
  ```
  sudo systemctl restart k3s.service
  ```
5. Copy kubeconfig to ~/.kube
  ```
  mkdir ~/.kube
  sudo cp /etc/rancher/k3s/k3s.yaml ~/.kube/config
  sudo chown $USER ~/.kube/config
  export KUBECONFIG=~/.kube/config
  ```
6. Generate cert-manager YAML files.
  ```
  helm template cert-manager ./cert-manager-v1.7.1.tgz --output-dir . \
      --namespace cert-manager \
      --set image.repository=myregistry.local:5000/quay.io/jetstack/cert-manager-controller \
      --set webhook.image.repository=myregistry.local:5000/quay.io/jetstack/cert-manager-webhook \
      --set cainjector.image.repository=myregistry.local:5000/quay.io/jetstack/cert-manager-cainjector \
      --set startupapicheck.image.repository=myregistry.local:5000/quay.io/jetstack/cert-manager-ctl
  ```
7. Install cert-manager.
  ```
  kubectl create namespace cert-manager
  kubectl apply -f cert-manager/cert-manager-crd.yaml
  kubectl apply -R -f ./cert-manager
  ```
8. Create `create-self-signed-cert.sh`, copy the content in the [signed certificate guide ](http://docs.rancher.cn/docs/rancher2.5/installation/resources/advanced/self-signed-ssl/_index/#4-%E5%A6%82%E4%BD%95%E7%94%9F%E6%88%90%E8%87%AA%E7%AD%BE%E5%90%8D%E8%AF%81%E4%B9%A6)
9. Make create-self-signed-cert.sh be excutable.
  ```
  chmod +x create-self-signed-cert.sh
  ```
10. Create certificate files for helm-install.local . Remember to change your private IP.
  ```
  ./create-self-signed-cert.sh --ssl-domain=helm-install.local --ssl-trusted-ip=192.168.0.50
  ```
11. Create cattle-system namespace
  ```
  kubectl create ns cattle-system
  ```
12. Create tls-sa secret.
  ```
  kubectl -n cattle-system create secret generic tls-ca \
    --from-file=cacerts.pem=./cacerts.pem
  ```
13. Create tls-rancher-ingress secret.
  ```
  kubectl -n cattle-system create secret tls tls-rancher-ingress \
    --cert=tls.crt \
    --key=tls.key
  ```
14. Generate Rancher YAML files.
  ```
  helm template rancher ./rancher-2.6.4-rc13.tgz --output-dir . \
      --no-hooks \
      --namespace cattle-system \
      --set hostname=helm-install.local \
      --set rancherImageTag=v2.6.4-rc13 \
      --set rancherImage=myregistry.local:5000/rancher/rancher \
      --set systemDefaultRegistry=myregistry.local:5000 \
      --set useBundledSystemChart=true \
      --set ingress.tls.source=secret \
      --set privateCA=true
  ```
15. Install Rancher 
  ```
  kubectl -n cattle-system apply -R -f ./rancher
  
  davidtclin@davidtclin-Standard-PC-Q35-ICH9-2009:~$ kubectl -n cattle-system apply -R -f ./rancher
  clusterrolebinding.rbac.authorization.k8s.io/rancher created
  deployment.apps/rancher created
  ingress.networking.k8s.io/rancher created
  service/rancher created 
  serviceaccount/rancher created
  
  ```
  ![image](https://user-images.githubusercontent.com/29251855/160362486-4e521a71-9b31-4460-a4a0-958653dbd56d.png)

#### Phase 6: Edit Rancher vm hosts file and boostrap Rancher
1. Edit the `/etc/hosts` file in Rancher virtual machine 
  ```
  192.168.0.50    myregistry.local
  ```
  ![image](https://user-images.githubusercontent.com/29251855/160388430-d91b7a6a-8f6b-486a-a07d-9f13dc5e8ece.png)

2. Edit the `/etc/hosts` file in Host machine
  ```
  192.168.0.50 helm-install.local
  ```
  ![image](https://user-images.githubusercontent.com/29251855/160389038-40cf1803-4914-4dae-83a7-0a547872429f.png)

3. Open browser on your host machine, access the URL https://helm-install.local/ 
  ![image](https://user-images.githubusercontent.com/29251855/160363090-3741cb41-a48e-4512-90ab-0ef2287e27ae.png)
 
4. Copy the the helm command and run in the rancher virtual machine terminal 

5. Get the password and complete the bootstrap process
   ![image](https://user-images.githubusercontent.com/29251855/160363211-980d17a8-8441-45ee-9af2-f4d000ac9c06.png)

#### Phase 7: Edit Harvester cluster hosts files
1. Edit `/etc/hosts` on harvester node machine 
  ```
  # vim /etc/hosts
  192.168.0.50 myregistry.local helm-install.local 
  ```
  ![image](https://user-images.githubusercontent.com/29251855/160390326-d89b5ece-2cab-48bc-a1ee-35fd7f8cc648.png)

2. Set registries.yaml in Harvester. /etc/rancher/rke2/registries.yaml
  ```
  # vim /etc/rancher/rke2/registries.yaml
  mirrors:
    docker.io:
      endpoint:
        - "https://myregistry.local:5000/"
  configs:
    "myregistry.local:5000":
      tls:
        insecure_skip_verify: true
  ```
3. Restart Harvester
  ```
  systemctl restart rke2-server.service
  ```

#### Phase 8: Add host mapping to coredns configmap and deployments 
1. Open K9s -> enter `: configmap` -> search dns
  ![image](https://user-images.githubusercontent.com/29251855/160373835-c4aa114e-d07d-4ec6-8ff7-ba742008bf9a.png)
2. Edit the `rke2-coredns-rke2-coredns`
3. Add the following content and save 
  ```
  hosts /etc/coredns/customdomains.db helm-install.local {\n
      \   fallthrough\n    }
  ```

  ```
  customdomains.db: |
      192.168.0.50 helm-install.local
  ```
  
  ```
  data:
    Corefile: ".:53 {\n    errors \n    health  {\n        lameduck 5s\n    }\n    ready
      \n    kubernetes   cluster.local  cluster.local in-addr.arpa ip6.arpa {\n        pods
      insecure\n        fallthrough in-addr.arpa ip6.arpa\n        ttl 30\n    }\n    prometheus
      \  0.0.0.0:9153\n   hosts /etc/coredns/customdomains.db helm-install.local {\n
      \   fallthrough\n    }\n forward   . /etc/resolv.conf\n    cache   30\n    loop
      \n    reload \n    loadbalance \n}"
    customdomains.db: |
      192.168.0.50 helm-install.local
  ```
  
4.  Open K9s -> enter `: deployment` -> search dns 
5.  Edit the rke2-coredns-rke2-coredns 
  ![image](https://user-images.githubusercontent.com/29251855/160374673-cadbf036-7a76-48c3-8029-306976f69ef0.png)
6.  Add the following content and save
  ```
  - key: customdomains.db
              path: customdomains.db
  ```
  ![image](https://user-images.githubusercontent.com/29251855/160374909-e8364442-9401-468f-8c3a-5f7ce9c1443b.png)


#### Phase 9: Import Harvester from Rancher UI
1. Environment preparation as above steps
2. Open Rancher virtualization management page and click import 
  ![image](https://user-images.githubusercontent.com/29251855/160376484-8d4503ab-32e3-4579-bb54-9f317763e795.png)

3. Copy the registration URL and paste in the Harvester cluster URL settings
  ![image](https://user-images.githubusercontent.com/29251855/160376541-8ad0a58b-e118-4748-9e2d-f0a99eeaca2a.png)

**Comment by TachunLin** [comment]
After we can import air gapped Rancher to Harvester. 
Then while we provisioning RKE2 it stuck in `Waitingfordatasecret` and can't create RKE2 cluster VM to Harvester. 
  ![image2](https://user-images.githubusercontent.com/29251855/160512547-ba1eaa44-df01-4faa-be77-5aba8beaf6f1.png)

We need to follow up to investigate this issue.

**Comment by noahgildersleeve** [comment]
Closing as fixed per comment in Harvester issue
https://github.com/harvester/harvester/issues/1808#issuecomment-1100025768
